### PR TITLE
feat: arrays support for setVariable

### LINKED ISSUE #31: Mocks: support arrays when using setVariable
**Is your feature request related to a problem? Please describe.**
Currently, an error is thrown when using `setVariable` in a variable of type array (any array)

**Describe the solution you'd like**
Arrays do have an special handling in the storage, we need to adapt.
As well, tests for multiple array types and structs with arrays should be added

**Additional context**
Information about how Hardhat handles arrays in the storage can be found [here](https://docs.soliditylang.org/en/v0.6.8/internals/layout_in_storage.html#mappings-and-dynamic-arrays)

There is an skipped test that should be unskipped as soon as this is working:
`Mock: Editable storage logic - should be able to set a uint256[] variable`

### PR DESCRIPTION
**Pull request by tonykogias**
**Description**
Add support for arrays on `setVariable` for Mock contracts.

**Metadata**

- Fixes #31

### CODE REVIEW

File: src/utils/storage.ts [hunk 1]
```diff
@@ -406,8 +406,64 @@ function encodeVariable(
     }
     return slots;
   } else if (variableType.encoding === 'dynamic_array') {
-    // TODO: add support for array types
-    throw new Error('array types not yet supported. Follow this issue for more info https://github.com/defi-wonderland/smock/issues/31');
+    if (variableType.base === undefined) {
+      // Should never happen in practice but required to maintain proper typing.
+      throw new Error(`variable is an array but has no base: ${variableType}`);
```
Comment (0xGorilla) [inline src/utils/storage.ts:411]: How is this fixable from a user standpoint?

File: test/contracts/mock/StorageGetter.sol [hunk 2]
```diff
@@ -33,6 +33,7 @@ contract StorageGetter {
   mapping(address => bool) _addressToBoolMap;
   mapping(address => address) _addressToAddressMap;
   uint256[] internal _uint256Array;
+  uint16[][] internal _uint2DArray;
```
Comment (wei3erHase) [inline test/contracts/mock/StorageGetter.sol:36]: blz try with `int16` and use negative numbers :)

File: test/unit/mock/editable-storage-logic.spec.ts [hunk 3]
```diff
@@ -149,9 +149,22 @@ describe('Mock: Editable storage logic', () => {
...
+      await mock.setVariable('_uint2DArray', [
+        [1, 2],
+        [3, 4],
+        [5, 6],
+      ]);
+      expect(await mock.getUint2D16Array()).to.eql([
+        [1, 2],
+        [3, 4],
+        [5, 6],
+      ]);
```
Comment (wei3erHase) [inline test/unit/mock/editable-storage-logic.spec.ts:167]: wowowowowowowo 🚀

File: src/utils/storage.ts [hunk 4]
```diff
@@ -406,8 +406,64 @@ function encodeVariable(
...
+    }
+    let slots: StorageSlotPair[] = [];
+    // In slotKey we save the length of the array
+    slots = slots.concat({
```
Comment (0xGorilla) [inline src/utils/storage.ts:415]: Let's add the object in initialisation:

```typescript
let slots: StorageSlotPair[] = [{
    key: slotKey,
    val: padNumHexSlotValue(variable.length, 0),
}];
```

File: src/utils/storage.ts [hunk 5]
```diff
@@ -406,8 +406,64 @@ function encodeVariable(
...
+    } else if (variableType.base.startsWith('t_uint') || variableType.base.startsWith('t_int')) {
+      // We find the number of bits from the base and divide it with 8 to get the number of bytes
+      numberOfBytes = Number(variableType.base.replace(/\D/g, '')) / 8;
+      // If we have more than 16Bytes for each value then we don't care about packed variables.
+      numberOfBytes > 16 ? 0 : numberOfBytes;
+    }
+
+    let offset: number = -numberOfBytes;
+    for (let i = 0; i < variable.length; i++) {
+      // If the values are pakced then we need to keep track of the offset and when to change slot
```
Comment (0xGorilla) [inline src/utils/storage.ts:437]:
```suggestion
      // If the values are packed then we need to keep track of the offset and when to change slot
```

**Review by 0xGorilla — CHANGES_REQUESTED** [review]
This is 🔥

File: test/unit/mock/editable-storage-logic.spec.ts [hunk 6]
```diff
@@ -149,9 +149,22 @@ describe('Mock: Editable storage logic', () => {
...
+      await mock.setVariable('_int2DArray', [
+        [-1, -2],
+        [-3, -4],
+        [5, -6],
+      ]);
+      expect(await mock.getInt2D16Array()).to.eql([
+        [-1, -2],
+        [-3, -4],
+        [5, -6],
+      ]);
```
Comment (0xGorilla) [inline test/unit/mock/editable-storage-logic.spec.ts:167]:
```suggestion
      const arr = [
        [-1, -2],
        [-3, -4],
        [5, -6],
      ];
      await mock.setVariable('_int2DArray', arr);
      expect(await mock.getInt2D16Array()).to.eql(arr);
```

**Review by 0xGorilla — APPROVED** [review]

**Comment by murderteeth** [comment]
thanks @tonykogias !

**Pull request merged**